### PR TITLE
PYTHON-2371 Add Azure and GCP support for CSFLE

### PR DESCRIPTION
**Pull request by prashantmital**
Depends on https://github.com/mongodb/libmongocrypt/pull/118

### CODE REVIEW

File: test/test_encryption.py [hunk 1]
```diff
@@ -1105,5 +1107,141 @@ def test_05_endpoint_invalid_host(self):
...
+    def read_extjson(self, filename):
+        with open(os.path.join(BASE, 'custom', filename), 'r') as fp:
+            return json_util.loads(fp.read(), json_options=JSON_OPTS)
+
+    def setup_keyvault(self, DEK):
+        keyvault = self.client.get_database(self.KEYVAULT_DB).get_collection(
+            self.KEYVAULT_COLL, codec_options=OPTS,
+            write_concern=WriteConcern('majority'))
+        keyvault.drop()
+        keyvault.insert_one(DEK)
```
Comment (ShaneHarvey) [inline test/test_encryption.py:1124]: We should use the existing `create_key_vault` helper function.

File: test/test_encryption.py [hunk 2]
```diff
@@ -1105,5 +1107,141 @@ def test_05_endpoint_invalid_host(self):
                 'aws', master_key=master_key)
 
 
+class AzureGCPEncryptionTestMixin(object):
+    KMS_PROVIDER_MAP = None
+    KEYVAULT_DB = 'keyvault'
+    KEYVAULT_COLL = 'datakeys'
+
+    def read_extjson(self, filename):
```
Comment (ShaneHarvey) [inline test/test_encryption.py:1115]: We should use the existing `json_data` helper function.

File: test/test_encryption.py [hunk 3]
```diff
@@ -1105,5 +1107,141 @@ def test_05_endpoint_invalid_host(self):
...
+AZURE_CREDS = {
+    'tenantId': os.environ.get('FLE_AZURE_TENANTID', ''),
+    'clientId': os.environ.get('FLE_AZURE_CLIENTID', ''),
+    'clientSecret': os.environ.get('FLE_AZURE_CLIENTSECRET', '')}
+
+
+class TestAzureEncryption(AzureGCPEncryptionTestMixin,
+                          EncryptionIntegrationTest):
+    @classmethod
+    @unittest.skipUnless(_HAVE_PYMONGOCRYPT, 'pymongocrypt is not installed')
```
Comment (ShaneHarvey) [inline test/test_encryption.py:1184]: This check already exists in EncryptionIntegrationTest so I think we can remove it.

File: test/test_encryption.py [hunk 4]
```diff
@@ -1105,5 +1107,141 @@ def test_05_endpoint_invalid_host(self):
...
+
+GCP_CREDS = {
+    'email': os.environ.get('FLE_GCP_EMAIL', ''),
+    'privateKey': os.environ.get('FLE_GCP_PRIVATEKEY', '')}
+
+
+class TestGCPEncryption(AzureGCPEncryptionTestMixin,
+                        EncryptionIntegrationTest):
+    @classmethod
+    @unittest.skipUnless(_HAVE_PYMONGOCRYPT, 'pymongocrypt is not installed')
```
Comment (ShaneHarvey) [inline test/test_encryption.py:1219]: This check already exists in EncryptionIntegrationTest so I think we can remove it.

File: test/test_encryption.py [hunk 5]
```diff
@@ -1105,5 +1107,141 @@ def test_05_endpoint_invalid_host(self):
...
+        keyvault = self.client.get_database(self.KEYVAULT_DB).get_collection(
+            self.KEYVAULT_COLL, codec_options=OPTS,
+            write_concern=WriteConcern('majority'))
+        keyvault.drop()
+        keyvault.insert_one(DEK)
+
+    def _test_explicit(self, expectation):
+        client_encryption = ClientEncryption(
+            self.KMS_PROVIDER_MAP,
+            '.'.join([self.KEYVAULT_DB, self.KEYVAULT_COLL]),
```
Comment (ShaneHarvey) [inline test/test_encryption.py:1129]: Suggest refactoring `'.'.join([self.KEYVAULT_DB, self.KEYVAULT_COLL])` into a variable to make this more readable.

File: test/test_encryption.py (hunk 2 above)
Comment (prashantmital) [reply test/test_encryption.py:1115]: I was already using this, just forgot to remove this helper. Done now.

**Comment by prashantmital** [comment]
Ready for another look @ShaneHarvey

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]
Can you document the new gcp and azure keys in ClientEncryption and AutoEncryptionOpts?

File: pymongo/encryption.py [hunk 6]
```diff
@@ -358,6 +358,17 @@ def __init__(self, kms_providers, key_vault_namespace, key_vault_client,
...
                 These are the AWS access key ID and AWS secret access key used
                 to generate KMS messages.
+              - `azure`: Map with "tenantId", "clientId", and "clientSecret" as
+                strings. Additionally, "identityPlatformEndpoint" may also be
+                specified as a string (defaults to 'login.microsoftonline.com').
+                These are the Azure Active Directory credentials used to
+                generate Azure Key Vault messages.
+              - `gcp`: Map with "email" as a string and "privateKey" as a
+                base64 encoded string or `bytes`. Python 2 users must specify
+                "privateKey" as a base64 encoded string. Additionally,
```
Comment (ShaneHarvey) [inline pymongo/encryption.py:368]: > Python 2 users must specify "privateKey" as a base64 encoded string.

Why? Can we make this work the same in Python 2 and 3?

Comment (prashantmital) [reply pymongo/encryption.py]: Changed. I will make the necessary code change in the PR to fix failing EVG tests.

**Review by ShaneHarvey — COMMENTED** [review]
Can you also update the "master_key" docs for ClientEncryption.create_data_key?

**Comment by prashantmital** [comment]
@ShaneHarvey updated.

File: test/test_encryption.py [hunk 7]
```diff
@@ -1105,5 +1107,130 @@ def test_05_endpoint_invalid_host(self):
...
+            self.assertEqual(output_doc[key], value)
+
+
+AZURE_CREDS = {
+    'tenantId': os.environ.get('FLE_AZURE_TENANTID', ''),
+    'clientId': os.environ.get('FLE_AZURE_CLIENTID', ''),
+    'clientSecret': os.environ.get('FLE_AZURE_CLIENTSECRET', '')}
+
+
+class TestAzureEncryption(AzureGCPEncryptionTestMixin,
```
Comment (ShaneHarvey) [inline test/test_encryption.py:1178]: How come these tests never run in evergreen? Does this need to be `TestAzureEncryption(EncryptionIntegrationTest, AzureGCPEncryptionTestMixin)`?

**Comment by ShaneHarvey** [comment]
I also opened https://jira.mongodb.org/browse/PYTHON-2405 to fix the `AttributeError: function/symbol 'mongocrypt_setopt_kms_providers' not found in library 'libmongocrypt.so.0':` test failures.

**Comment by prashantmital** [comment]
Finally figured out why the tests weren't running - I wasn't exporting the environment variable names correctly.

**Comment by ShaneHarvey** [comment]
Can you schedule the latest patch to run python 2.7-3.6 to make sure the new tests work there too?

File: .evergreen/run-tests.sh [hunk 8]
```diff
@@ -47,6 +47,9 @@ if [ "$SSL" != "nossl" ]; then
     export CA_PEM="$DRIVERS_TOOLS/.evergreen/x509gen/ca.pem"
 fi
 
+# For createvirtualenv.
+. .evergreen/utils.sh
```
Comment (ShaneHarvey) [inline .evergreen/run-tests.sh:51]: Can you revert the virtualenv changes? We can do https://jira.mongodb.org/browse/PYTHON-1960 and (https://jira.mongodb.org/browse/PYTHON-2405) in a new PR. That way we won't drag down this PR.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: .evergreen/run-tests.sh [hunk 9]
```diff
@@ -152,7 +152,7 @@ if [ -n "$TEST_ENCRYPTION" ]; then
         exit 1
     fi
 
-    git clone --branch master git@github.com:mongodb/libmongocrypt.git libmongocrypt_git
+    git clone --branch PYTHON-2409/add-test-local-kms-python-2 git@github.com:prashantmital/libmongocrypt.git libmongocrypt_git
```
Comment (ShaneHarvey) [inline .evergreen/run-tests.sh:155]: I think we no longer need this.

File: pymongo/encryption.py [hunk 10]
```diff
@@ -358,6 +358,16 @@ def __init__(self, kms_providers, key_vault_namespace, key_vault_client,
               - `aws`: Map with "accessKeyId" and "secretAccessKey" as strings.
                 These are the AWS access key ID and AWS secret access key used
                 to generate KMS messages.
+              - `azure`: Map with "tenantId", "clientId", and "clientSecret" as
+                strings. Additionally, "identityPlatformEndpoint" may also be
+                specified as a string (defaults to 'login.microsoftonline.com').
+                These are the Azure Active Directory credentials used to
+                generate Azure Key Vault messages.
+              - `gcp`: Map with "email" as a string and "privateKey" as a
+                base64 encoded string or `bytes`. Additionally,
```
Comment (ShaneHarvey) [inline pymongo/encryption.py:367]: > "privateKey" as a base64 encoded string or `bytes`.

What about:

```
"privateKey" as `bytes` or a base64 encoded string (unicode on Python 2).
```

Note we should keep these docs in sync with `AutoEncryptionOpts` below.

File: pymongo/encryption_options.py [hunk 11]
```diff
@@ -59,6 +59,17 @@ def __init__(self, kms_providers, key_vault_namespace,
...
                 These are the AWS access key ID and AWS secret access key used
                 to generate KMS messages.
+              - `azure`: Map with "tenantId", "clientId", and "clientSecret" as
+                strings. Additionally, "identityPlatformEndpoint" may also be
+                specified as a string (defaults to 'login.microsoftonline.com').
+                These are the Azure Active Directory credentials used to
+                generate Azure Key Vault messages.
+              - `gcp`: Map with "email" as a string and "privateKey" as a
+                base64 encoded string or `bytes`. Python 2 users must specify
+                "privateKey" as a base64 encoded string. Additionally,
```
Comment (ShaneHarvey) [inline pymongo/encryption_options.py:69]: > Python 2 users must specify "privateKey" as a base64 encoded string

This isn't true at least not anymore. I suggest:
```
"privateKey" as `bytes` or a base64 encoded string (unicode on Python 2).
```

File: pymongo/encryption.py [hunk 12]
```diff
@@ -358,6 +358,16 @@ def __init__(self, kms_providers, key_vault_namespace, key_vault_client,
...
+                "endpoint" may also be specified as a string (defaults to
+                'oauth2.googleapis.com'). These are the credentials used to
+                generate Google Cloud KMS messages.
               - `local`: Map with "key" as a 96-byte array or string. "key"
```
Comment (ShaneHarvey) [inline pymongo/encryption.py:371]: We should mention base64 encoding here too (and in AutoEncryptionOpts):
```
"key" as `bytes` (96 bytes in length) or a base64 encoded string (unicode on Python 2) which decodes to 96 bytes.
```

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: test/test_encryption.py [hunk 13]
```diff
@@ -1105,5 +1108,130 @@ def test_05_endpoint_invalid_host(self):
...
+    KEYVAULT_COLL = 'datakeys'
+
+    def setUp(self):
+        keyvault = self.client.get_database(
+            self.KEYVAULT_DB).get_collection(
+            self.KEYVAULT_COLL)
+        create_key_vault(keyvault, self.DEK)
+
+    def _test_explicit(self, expectation):
+        client_encryption = ClientEncryption(
```
Comment (ShaneHarvey) [inline test/test_encryption.py:1124]: ClientEncryption needs to be explicitly closed.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: pymongo/encryption.py (hunk 10 above)
Comment (prashantmital) [reply pymongo/encryption.py]: Done.

File: pymongo/encryption.py (hunk 12 above)
Comment (prashantmital) [reply pymongo/encryption.py]: Done.

File: pymongo/encryption_options.py (hunk 11 above)
Comment (prashantmital) [reply pymongo/encryption_options.py]: Done.

File: test/test_encryption.py (hunk 13 above)
Comment (prashantmital) [reply test/test_encryption.py:1124]: Done.

**Review by ShaneHarvey — APPROVED** [review]
LGTM!

**Pull request merged**